File: src/ClientModel/ClientModelProperty.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// A property that exists within a model defined by the client.
    /// </summary>
    public class ClientModelProperty
    {
        /// <summary>
        /// Create a new ClientModelProperty with the provided properties.
        /// </summary>
        /// <param name="name">The name of this property.</param>
        /// <param name="description">The description of this property.</param>
        /// <param name="annotationArguments">The arguments that go into this property's JsonProperty annotation.</param>
        /// <param name="isXmlAttribute">Whether or not this property is an attribute when serialized to XML.</param>
        /// <param name="xmlName">This property's name when serialized to XML.</param>
        /// <param name="serializedName">This property's name when it is serialized.</param>
        /// <param name="isXmlWrapper">Whether or not this property is a container.</param>
        /// <param name="xmlListElementName">The name of each list element tag within an XML list property.</param>
        /// <param name="wireType">The type of this property as it is transmitted across the network (across the wire).</param>
        /// <param name="clientType">The type of this property as it will be exposed via the client.</param>
        /// <param name="isConstant">Whether or not this property has a constant value.</param>
        /// <param name="defaultValue">The default value expression of this property.</param>
        /// <param name="isReadOnly">Whether or not this property's value can be changed by the client library.</param>
        /// <param name="headerCollectionPrefix">The prefix of the headers that make up this property's values.</param>
        public ClientModelProperty(string name, string description, string annotationArguments, bool isXmlAttribute, string xmlName, string serializedName, bool isXmlWrapper, string xmlListElementName, IType wireType, IType clientType, bool isConstant, string defaultValue, bool isReadOnly, bool wasFlattened, string headerCollectionPrefix)
        {
            Name = name;
            Description = description;
            AnnotationArguments = annotationArguments;
            IsXmlAttribute = isXmlAttribute;
            XmlName = xmlName;
            SerializedName = serializedName;
            IsXmlWrapper = isXmlWrapper;
            XmlListElementName = xmlListElementName;
            WireType = wireType;
            ClientType = clientType;
            IsConstant = isConstant;
            DefaultValue = defaultValue;
            IsReadOnly = isReadOnly;
            WasFlattened = wasFlattened;
            HeaderCollectionPrefix = headerCollectionPrefix;
        }

        /// <summary>
        /// Get the name of this property.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// Get the description of this property.
        /// </summary>
        public string Description { get; }

        /// <summary>
        /// Get the arguments that go into this property's JsonProperty annotation.
        /// </summary>
        public string AnnotationArguments { get; }

        /// <summary>
        /// Get whether or not this property is an attribute when serialized to XML.
        /// </summary>
        public bool IsXmlAttribute { get; }

        /// <summary>
        /// Get this property's name when serialized to XML.
        /// </summary>
        public string XmlName { get; }

        /// <summary>
        /// Get this property's name when it is serialized.
        /// </summary>
        public string SerializedName { get; }

        /// <summary>
        /// Get whether or not this property is a container.
        /// </summary>
        public bool IsXmlWrapper { get; }

        /// <summary>
        /// The name of each list element tag within an XML list property.
        /// </summary>
        public string XmlListElementName { get; }

        /// <summary>
        /// The type of this property as it is transmitted across the network (across the wire).
        /// </summary>
        public IType WireType { get; }

        /// <summary>
        /// The type of this property as it will be exposed via the client.
        /// </summary>
        public IType ClientType { get; }

        /// <summary>
        /// Get whether or not this property has a constant value.
        /// </summary>
        public bool IsConstant { get; }

        /// <summary>
        /// Get the default value expression of this property.
        /// </summary>
        public string DefaultValue { get; }

        /// <summary>
        /// Get whether or not this property's value can be changed by the client library.
        /// </summary>
        public bool IsReadOnly { get; }

        /// <summary>
        /// Whether or not this property was flattened.
        /// </summary>
        public bool WasFlattened { get; }

        /// <summary>
        /// The prefix of the headers that make up this property's values.
        /// </summary>
        public string HeaderCollectionPrefix { get; }

        /// <summary>
        /// Add this ServiceModelProperty's imports to the provided ISet of imports.
        /// </summary>
        /// <param name="imports">The set of imports to add to.</param>
        /// <param name="settings">The settings for this Java generator session.</param>
        public void AddImportsTo(ISet<string> imports, JavaSettings settings)
        {
            if (!string.IsNullOrEmpty(HeaderCollectionPrefix))
            {
                imports.Add("com.azure.common.annotations.HeaderCollection");
            }

            WireType.AddImportsTo(imports, false);
            ClientType.AddImportsTo(imports, false);

            if (WasFlattened)
            {
                imports.Add("com.azure.common.implementation.serializer.JsonFlatten");
            }

            if (settings.ShouldGenerateXmlSerialization)
            {
                imports.Add("com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement");
                imports.Add("com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty");
            }
            else
            {
                imports.Add("com.fasterxml.jackson.annotation.JsonProperty");
            }
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/clientmodel/ClientModelProperty.java
package com.azure.autorest.model.clientmodel;

import java.util.Set;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


/** 
 A property that exists within a model defined by the client.
*/
public class ClientModelProperty
{
    /**
     Create a new ClientModelProperty with the provided properties.

     @param name The name of this property.
     @param description The description of this property.
     @param annotationArguments The arguments that go into this property's JsonProperty annotation.
     @param isXmlAttribute Whether or not this property is an attribute when serialized to XML.
     @param xmlName This property's name when serialized to XML.
     @param serializedName This property's name when it is serialized.
     @param isXmlWrapper Whether or not this property is a container.
     @param xmlListElementName The name of each list element tag within an XML list property.
     @param wireType The type of this property as it is transmitted across the network (across the wire).
     @param clientType The type of this property as it will be exposed via the client.
     @param isConstant Whether or not this property has a constant value.
     @param defaultValue The default value expression of this property.
     @param isReadOnly Whether or not this property's value can be changed by the client library.
     @param headerCollectionPrefix The prefix of the headers that make up this property's values.
    */
    public ClientModelProperty(String name, String description, String annotationArguments, boolean isXmlAttribute, String xmlName, String serializedName, boolean isXmlWrapper, String xmlListElementName, IType wireType, IType clientType, boolean isConstant, String defaultValue, boolean isReadOnly, boolean wasFlattened, String headerCollectionPrefix)
    {
        Name = name;
        Description = description;
        AnnotationArguments = annotationArguments;
        IsXmlAttribute = isXmlAttribute;
        XmlName = xmlName;
        SerializedName = serializedName;
        IsXmlWrapper = isXmlWrapper;
        XmlListElementName = xmlListElementName;
        WireType = wireType;
        ClientType = clientType;
        IsConstant = isConstant;
        DefaultValue = defaultValue;
        IsReadOnly = isReadOnly;
        WasFlattened = wasFlattened;
        HeaderCollectionPrefix = headerCollectionPrefix;
    }

    /**
     Get the name of this property.
    */
    private String Name;
    public final String getName()
    {
        return Name;
    }

    /**
     Get the description of this property.
    */
    private String Description;
    public final String getDescription()
    {
        return Description;
    }

    /**
     Get the arguments that go into this property's JsonProperty annotation.
    */
    private String AnnotationArguments;
    public final String getAnnotationArguments()
    {
        return AnnotationArguments;
    }

    /**
     Get whether or not this property is an attribute when serialized to XML.
    */
    private boolean IsXmlAttribute;
    public final boolean getIsXmlAttribute()
    {
        return IsXmlAttribute;
    }

    /**
     Get this property's name when serialized to XML.
    */
    private String XmlName;
    public final String getXmlName()
    {
        return XmlName;
    }

    /**
     Get this property's name when it is serialized.
    */
    private String SerializedName;
    public final String getSerializedName()
    {
        return SerializedName;
    }

    /**
     Get whether or not this property is a container.
    */
    private boolean IsXmlWrapper;
    public final boolean getIsXmlWrapper()
    {
        return IsXmlWrapper;
    }

    /**
     The name of each list element tag within an XML list property.
    */
    private String XmlListElementName;
    public final String getXmlListElementName()
    {
        return XmlListElementName;
    }

    /**
     The type of this property as it is transmitted across the network (across the wire).
    */
    private IType WireType;
    public final IType getWireType()
    {
        return WireType;
    }

    /**
     The type of this property as it will be exposed via the client.
    */
    private IType ClientType;
    public final IType getClientType()
    {
        return ClientType;
    }

    /**
     Get whether or not this property has a constant value.
    */
    private boolean IsConstant;
    public final boolean getIsConstant()
    {
        return IsConstant;
    }

    /**
     Get the default value expression of this property.
    */
    private String DefaultValue;
    public final String getDefaultValue()
    {
        return DefaultValue;
    }

    /**
     Get whether or not this property's value can be changed by the client library.
    */
    private boolean IsReadOnly;
    public final boolean getIsReadOnly()
    {
        return IsReadOnly;
    }

    /**
     Whether or not this property was flattened.
    */
    private boolean WasFlattened;
    public final boolean getWasFlattened()
    {
        return WasFlattened;
    }

    /**
     The prefix of the headers that make up this property's values.
    */
    private String HeaderCollectionPrefix;
    public final String getHeaderCollectionPrefix()
    {
        return HeaderCollectionPrefix;
    }

    /**
     Add this ServiceModelProperty's imports to the provided ISet of imports.

     @param imports The set of imports to add to.
    */
    public final void AddImportsTo(Set<String> imports, boolean shouldGenerateXmlSerialization)
    {
        if (getHeaderCollectionPrefix() != null && !getHeaderCollectionPrefix().isEmpty())
        {
            imports.add("com.azure.core.implementation.annotation.HeaderCollection");
        }

        getWireType().AddImportsTo(imports, false);
        getClientType().AddImportsTo(imports, false);

        if (getClientType().equals(ArrayType.ByteArray))
        {
            imports.add("com.azure.core.implementation.util.ImplUtils");
        }

        if (getWasFlattened())
        {
            imports.add("com.azure.core.implementation.annotation.JsonFlatten");
        }

        if (shouldGenerateXmlSerialization)
        {
            imports.add("com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement");
            if (getIsXmlWrapper())
            {
                imports.add("com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty");
            }
        }
        else
        {
            imports.add("com.fasterxml.jackson.annotation.JsonProperty");
        }
    }
}